src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/DirectoryBasedStoreProvider.cs

```diff
@@ -0,0 +1,454 @@
+                {
+                    if (_readOnly)
+                    {
+                        // Windows compatibility, the readonly check isn't done until after a match is found.
+                        throw new CryptographicException(SR.Cryptography_X509_StoreReadOnly);
+                    }
+
+                    File.Delete(currentFilename);
+                }
+            }
+
+            // Null out _certificates so the next call to get_Certificates causes a re-scan.
+            _certificates = null;
+        }
+
+        private static string FindExistingFilename(X509Certificate2 cert, string storePath, out bool hadCandidates)
+        {
+            hadCandidates = false;
+
+            foreach (string maybeMatch in Directory.EnumerateFiles(storePath, cert.Thumbprint + PfxWildcard))
+            {
+                hadCandidates = true;
+
+                try
+                {
+                    using (X509Certificate2 candidate = new X509Certificate2(maybeMatch))
+                    {
+                        if (candidate.Equals(cert))
+                        {
+                            return maybeMatch;
+                        }
+                    }
+                }
+                catch (CryptographicException)
+                {
+                    // Contents weren't interpretable as a certificate, so it's not a match.
+                }
+            }
+
+            return null;
+        }
+
+        private string FindOpenSlot(string thumbprint)
+        {
+            // We already know that {thumbprint}.pfx is taken, so start with {thumbprint}.1.pfx
+
+            // We need space for {thumbprint} (thumbprint.Length)
+            // And ".0.pfx" (6)
+            // If MaxSaveAttempts is big enough to use more than one digit, we need that space, too (MaxSaveAttempts / 10)
+            StringBuilder pathBuilder = new StringBuilder(thumbprint.Length + PfxOrdinalWildcard.Length + (MaxSaveAttempts / 10));
+            HashSet<string> existingFiles = new HashSet<string>(Directory.EnumerateFiles(_storePath, thumbprint + PfxOrdinalWildcard));
+
+            for (int i = 1; i <= MaxSaveAttempts; i++)
+            {
+                pathBuilder.Clear();
+
+                pathBuilder.Append(thumbprint);
+                pathBuilder.Append('.');
+                pathBuilder.Append(i);
+                pathBuilder.Append(PfxExtension);
+
+                string builtPath = pathBuilder.ToString();
+
+                if (!existingFiles.Contains(builtPath))
+                {
+                    return Path.Combine(_storePath, builtPath);
+                }
+            }
+
+            throw new CryptographicException(SR.Cryptography_X509_StoreNoFileAvailable);
+        }
+
+        private void WatcherError(object sender, ErrorEventArgs e)
+        {
+            FlushCache();
+        }
+
+        private void DirectoryChanged(object sender, FileSystemEventArgs e)
+        {
+            FlushCache();
+        }
```

Comment:
Since you don't need the strongly typed event args, you can collapse both WatcherError and DirectoryChanged into a single:
```C#
private void WatcherChanged(object sender, EventArgs e)
{
    FlushCache();
}
```
and then use that one method for all of the callbacks.  Delegates support contravariance in this regard.